File: Assignment 1/Problem 1/src/brute_force.f90
! @Author: Gilbert Young
! @Time: 2024/9/13 21:00
! @File_name: brute_force.f90
! @IDE: Vscode
! @Formatter: Fprettify
! @Description: This script performs a brute-force search for solutions to the Diophantine equation a^5 + b^5 + c^5 + d^5 = e^5.

program diophantine_fifths
    implicit none
    integer, parameter :: N = 200  ! Upper bound
    integer(kind=8) :: a, b, c, d, e
    real(kind=8) :: start_time, end_time
    character(len=200) :: output

    ! Start timing
    call cpu_time(start_time)

    ! Brute-force search for solutions
    do a = 1, N
        ! Ensure b starts from a to avoid duplicate solutions (a <= b)
        do b = a, N
            do c = b, N
                do d = c, N
                    do e = d, N
                        ! Check if a^5 + b^5 + c^5 + d^5 == e^5
                        if (a**5 + b**5 + c**5 + d**5 == e**5) then
                            ! Format and store the solution in output string
                            write (output, '(A, I0, A, I0, A, I0, A, I0, A, I0, A)') &
                                'Solution:', a, '^5+', b, '^5+', c, '^5+', d, '^5=', e, '^5'
                            ! Print the formatted solution
                            print *, trim(output)
                        end if
                    end do
                end do
            end do
        end do
    end do

    ! End timing and print elapsed time
    call cpu_time(end_time)
    print *, 'Elapsed time for brute_force:', end_time - start_time, 'seconds'
end program diophantine_fifths

File: Assignment_1/Problem_1/src/brute_force.py
"""
@Author: Gilbert Young
@Time: 2024/9/13 21:00
@File_name: brute_force.py
@IDE: Vscode
@Formatter: Black
@Description: This script performs a brute-force search for solutions to the Diophantine equation a^5 + b^5 + c^5 + d^5 = e^5.
"""

import time


def find_diophantine_solution(limit):
    # List of n to n^5; Dictionary of n^5 to n
    pow_5 = [n**5 for n in range(limit + 1)]
    pow5_to_n = {n**5: n for n in range(limit + 1)}

    # Outer four nested loops to find the sum of the fifth powers of four numbers
    for a in range(1, limit + 1):
        for b in range(a, limit + 1):
            for c in range(b, limit + 1):
                for d in range(c, limit + 1):
                    # Compute the sum of the fifth powers of the first four numbers
                    pow_5_sum = pow_5[a] + pow_5[b] + pow_5[c] + pow_5[d]
                    # Look for pow_5_sum in the fifth power dictionary
                    if pow_5_sum in pow5_to_n:
                        e = pow5_to_n[pow_5_sum]
                        return (a, b, c, d, e)

    return None


start_time = time.time()
# Set the search range
limit = 200
# Call the function to find the solution
result = find_diophantine_solution(limit)
# Measure elapsed time
total_time = time.time() - start_time

# Output the result and time
if result:
    a, b, c, d, e = result
    print(f"Solution: {a}^5 + {b}^5 + {c}^5 + {d}^5 = {e}^5")
else:
    print("No solution found")
print(f"Total time: {total_time:.6f} seconds")
